### PR TITLE
Implement Particle Swarm Optimization (PSO) for Swarm Class

### LINKED ISSUE #15: Enhancing PSO: Updates, Modifications, and Completion
### Title:
Implement Particle Swarm Optimization (PSO) for Swarm Class Initialization, Optimization, and Updation

#### Issue Description:
The task involves implementing Particle Swarm Optimization (PSO) functionality for the Swarm class. This includes initialization of individuals, optimization of the swarm, and updating velocities and weights of the individuals within the swarm.

#### Proposed Changes:
1. **Initialization of Swarm Individuals:**
   - Implement a function to initialize all individuals of the swarm with the provided data.
   - Extract input features and target labels from the data.
   - Initialize individuals with random weights and biases.
   - Initialize velocities of individuals.
   - Initialize local best individuals as a deep copy of the individuals.
   - Find the global best individual.

2. **Optimization of the Swarm:**
   - Implement a function to execute the optimization algorithm until the maximum number of iterations is reached.
   - Update the swarm by iteratively updating velocities and weights of individuals.
   - Calculate the fitness of individuals.
   - Update local best and global best individuals.

3. **Updating Velocities and Weights:**
   - Implement a function to update the velocities and weights of the swarm.
   - Calculate fitness and update local best and global best fitness.

#### Expected Outcomes:
- Improved functionality of the Swarm class by integrating PSO capabilities.
- Efficient optimization of weights and biases for various applications, such as neural network training and optimization problems.

#### Additional Information:
This issue is essential for enhancing the functionality of the Swarm class by incorporating PSO capabilities. Implementing PSO will enable the class to efficiently optimize weights and biases for various applications, such as neural network training and optimization problems.

### PR DESCRIPTION
**Pull request by avneesh777**
#### Issue Description:
This pull request fixes #15  and addresses the implementation of Particle Swarm Optimization (PSO) functionality for the Swarm class, as outlined in the issue.

#### Changes Made:
1. **Initialization of Swarm Individuals:**
   - Implemented a function to initialize all individuals of the swarm with the provided data.
   - Extracted input features and target labels from the data.
   - Initialized individuals with random weights and biases.
   - Initialized velocities of individuals.
   - Initialized local best individuals as a deep copy of the individuals.
   - Found the global best individual.

2. **Optimization of the Swarm:**
   - Implemented a function to execute the optimization algorithm until the maximum number of iterations is reached.
   - Updated the swarm by iteratively updating velocities and weights of individuals.
   - Calculated the fitness of individuals.
   - Updated local best and global best individuals.

3. **Updating Velocities and Weights:**
   - Implemented a function to update the velocities and weights of the swarm.
   - Calculated fitness and updated local best and global best fitness.

#### Expected Outcomes:
- Improved functionality of the Swarm class by integrating PSO capabilities.
- Efficient optimization of weights and biases for various applications, such as neural network training and optimization problems.

### CODE REVIEW

File: src/pages/3_PSO.py [hunk 1]
```diff
@@ -45,8 +45,281 @@
     6. Retrieve the best solution found by the PSO algorithm from the swarm.
 
     ## Source code explanation
+    
+    The `sigmoid` function applies the sigmoid function to the given input array or scalar.
+    **Inputs:**
+        - X: Input data or values.
```
Comment (Haleshot) [inline src/pages/3_PSO.py:51]: Please look into the formatting of the markdown syntax as it appears on the output. Leave a space for it to be shown correctly as done in the later half of the program.

**Review by Haleshot — COMMENTED** [review]
Please look into the formatting of the markdown syntax as it appears on the output. Leave a space for it to be shown correctly as done in the later half of the program implementation. The inputs and functionality parts don't appear in the right formatted order.

**Review by Haleshot — CHANGES_REQUESTED** [review]
Overall formatting needs to be done with respect to Input points, functionality points. Those are the main headings with pointers coming below them.

**Comment by avneesh777** [comment]
changes has been done as per your instructions please check it once.

File: src/pages/3_PSO.py [hunk 2]
```diff
@@ -45,8 +45,284 @@
...
+    The `sigmoid` function applies the sigmoid function to the given input array or scalar.
+    
+    - **Inputs:**
+        - X: Input data or values.
+    
+    - **Functionality:**
+        - Applies the hyperbolic tangent (tanh) activation function element-wise to the input array or scalar.
+    
+    ```python
+    def sigmoid(Z):
```
Comment (Haleshot) [inline src/pages/3_PSO.py:58]: This function doesn't contain the docstrings (''') like the other functions like tanh. Please resolve.

File: src/pages/3_PSO.py [hunk 3]
```diff
@@ -45,8 +45,284 @@
...
+    - **Functionality:**
+            - Optimizes the swarm by updating velocities and weights.
+            - Updates local best and global best individuals iteratively.
+            - Terminates when the maximum number of iterations is reached.
+    
+    **Returns:**
+            Particle: The global best particle.
+            
+    ```python
+    def optimize(self):
```
Comment (Haleshot) [inline src/pages/3_PSO.py:239]: Where is the docstring commented format for this function like you mentioned for others.

**Review by Haleshot — CHANGES_REQUESTED** [review]
See comments mentioned in the specific lines and resolve.

File: src/pages/3_PSO.py [hunk 4]
```diff
@@ -45,8 +45,285 @@
...
 
     ## Source code explanation
+    
+    The `sigmoid` function applies the sigmoid function to the given input array or scalar.
+    
+    - **Inputs:**
+        - X: Input data or values.
+    
+    - **Functionality:**
+        - Applies the hyperbolic tangent (tanh) activation function element-wise to the input array or scalar.
```
Comment (Haleshot) [inline src/pages/3_PSO.py:55]: Sigmoid not tanh

**Comment by avneesh777** [comment]
Please verify that all of the changes you requested have been made.

**Review by Haleshot — APPROVED** [review]
LGTM!

**Pull request merged**